### PR TITLE
net: lwm2m: Fix SenML CBOR object link encoding

### LINKED ISSUE #52527: net: lwm2m: wrong SenML CBOR object link encoding
Zephyr's SenML CBOR encoding is not in line with the specification. It it using TLV format (encoding the object link into a 32 bit value) instead of using the text format (encoding the link to a string).

References:
- [JSON Format](https://www.openmobilealliance.org/release/LightweightM2M/V1_1-20180710-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1-20180710-A.html#Table-744-1-JSON-format-and-description)
- [Object Link Format](https://www.openmobilealliance.org/release/LightweightM2M/V1_1-20180710-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1-20180710-A.html#Table-C-1-Data-Types)
- [CBOR Example](https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/484#issuecomment-705490737)

### PR DESCRIPTION
**Pull request by broglep-work**
The previous patch could not be applied, so I had to manually re-apply the patch. I kept generated code & modification of it as separate commits for now to ease review, will squash when reviewed. 

further, clang-format has been applied to patch commit (99f2f689a82a72f3327ff8a9051cc4a3f3a8920a) 

fixes #52527

### CODE REVIEW

**Comment by rlubos** [comment]
CC @VeijoPesonen

**Review by rlubos — COMMENTED** [review]
Thank you @broglep-work for the contribution. The changes look reasonable to me, however I'd like @VeijoPesonen to review the code as he's the original SenML CBOR contributor.

However, it seems that the patch file is not complete? We still hit some compliance issues with the autogenerated files.

**Comment by VeijoPesonen** [comment]
@broglep-work Thanks, I'm in the middle of checking the PR but looks really good thus far. In the meanwhile would you please address the whitespace issues compliance check reports.

**Review by VeijoPesonen — COMMENTED** [review]
Looks good to me. @broglep-work Greatly appreciated :) I'll approve once you have fixed the minor whitespace issues.

**Comment by VeijoPesonen** [comment]
@broglep-work As the compliance check tells please provide a commit body on each separate commit message to describe the change a bit more.

**Comment by broglep-work** [comment]
Any suggestions how to fix those compliance check? 

I think it does not make much sense manually modifying patch files to make checkpatch happy, nor it does seam that there is an easy way to ignore that particular file. And trying to reverse engineer & modify the original files so that the patch file makes checkpatch happy feels worthless.

And there is another warning about max line length in the generated code, where clang-format is happy with exceeding 100 chars but checkpatch is not. As this is generated code, I'd also rather ignore those checks, or fix zcbor so that it does not produces lines longer than 100, instead of changing that code manually (which then causes potential future patch applys too fail)

**Comment by rlubos** [comment]
I had the impression that the whole idea for the patch file was to summarize the changes needed to align the autogenerated file with Zephyr requirements? @VeijoPesonen How was this handled in the past? @oyvindronningstad Perhaps you could chip in given you've been involved in the process in the past?

> And there is another warning about max line length in the generated code, where clang-format is happy with exceeding 100 chars but checkpatch is not.

Is that really the case for clang-format? There is an explicit limit set in the config file to 100 chars/line:
https://github.com/zephyrproject-rtos/zephyr/blob/main/.clang-format#L30

**Comment by broglep-work** [comment]
Not sure how well that patch approach worked in the past, but for my work it was not of much use. A document that described which high level modifications need to be applied to the generated would maybe have been more helpful. (Such as replace `9223372036854775807L` with `INT64_MIN`, Remove hardcoded limit of 99 with config, etc.)

My best idea at the moment is move all that code generation stuff outside of the current directory into a dedicated one and ignore that directory for checkpatch

> Is that really the case for clang-format? There is an explicit limit set in the config file to 100 chars/line:

As far as I know clang-format deviates from that 100 char limit in sensible cases, which in this case it seems to be good judgement of clang-format, re-formatting that few lines to have less than 100 lines makes the already hard to read code (because autogenerated) even more difficult to read.

Ideally that auto-generated file could also be excluded from checkpatch. If not, would probably be okish to modify the file manually so that we don't exceed 100 lines, but this will make the patch file approach of applying changes even more brittle than it already is

**Comment by broglep-work** [comment]
After many hours wasted due to checkpatch (and failure to tweak clang-format to respect max line length at all costs) , I think I've come to a solution that is as good I can get it.  

At least we now have much better approach for the code transformation using http://comby.dev which documents the transformations and should be more robust for the future. 

It is also easier to reason about the changes to the generated code. At least while writing the transformation I had several thought that the transformation could be maybe be better (thought that did not cross my mind I was when working with the patch file and reverse-engineer the intended transformations from the patch file). In some places where `ZCBOR_ARRAY_SIZE()` is used, we could probably just use `CONFIG_LWM2M_RW_SENML_CBOR_RECORDS`

File: subsys/net/lib/lwm2m/CMakeLists.txt [hunk 1]
```diff
@@ -71,8 +71,8 @@ zephyr_library_sources_ifdef(CONFIG_LWM2M_RW_CBOR_SUPPORT
 # SENML CBOR support
 zephyr_library_sources_ifdef(CONFIG_LWM2M_RW_SENML_CBOR_SUPPORT
     lwm2m_rw_senml_cbor.c
-    lwm2m_senml_cbor_decode.c
-    lwm2m_senml_cbor_encode.c
+        generated/lwm2m_senml_cbor_decode.c
+        generated/lwm2m_senml_cbor_encode.c
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/CMakeLists.txt:75]: I don't think an extra indent is needed here.

**Comment by rlubos** [comment]
Adding more reviewers into the loop given the PR now references new util for autogenerated ode tranformation. I'm ok with this, but wanted to hear from others.

**Comment by VeijoPesonen** [comment]
Placing the generated files to a separate folder and making Git to ignore it sounds reasonable for me. In my opinion introducing one more external tool like comby is no go but that's not for me to decide. Looks like a handy tool though.

**Comment by SeppoTakalo** [comment]
I'm not familiar with this tool `comby`

Is it really so much better that we should convert this specific converter to use it while nobody else in Zephyr do not.
It does not even appear in common repositories for common Linux distros.

Why not try to do the same transformation with a bit familiar tool that already has use in Zephyr: https://docs.zephyrproject.org/latest/develop/tools/coccinelle.html

**Comment by broglep-work** [comment]
As comby is not directly used as part of the regular code workflow, I don't see much of an issue. We already depend on zcbor python package as an external tool. 

A nice solution would be to integrate the code generation and transformation into cmake, in that case I agree, we should more carefully consider which tools are in use (and make sure that the get properly installed as part of the regular developer setup)

As long as we have a workflow to run a script occasionally, generate some code, that we then review & commit, I think we should be able to use tools that get the job done quickly. Comby seemed to be a popular choice, although I spent maybe 15 minutes of research. There might be other tools out there that are more in line with the technologies/ecosystem of zephyr

File: subsys/net/lib/lwm2m/CMakeLists.txt (hunk 1 above)
Comment (broglep-work) [reply subsys/net/lib/lwm2m/CMakeLists.txt]: missed that one, have fixed it (was introduced refactoring / auto-format)

**Comment by broglep-work** [comment]
> Why not try to do the same transformation with a bit familiar tool that already has use in Zephyr: https://docs.zephyrproject.org/latest/develop/tools/coccinelle.html

sounds reasonable, anyone here that has experience with that tool that can try convert from comby to coccinelle?

**Comment by carlescufi** [comment]
@broglep-work thanks very much for the proposal. I have added the TSC label so that we can discuss the approach there. If you'd like to join today, here are the details:
https://github.com/zephyrproject-rtos/zephyr/wiki/Technical-Steering-Committee-%28TSC%29

**Review by SeppoTakalo — DISMISSED** [review]
After a second thought, I'm proposing that we accept this patch as it is.

Yes, I would like to see the formatting done with Coccinelle, but we should open a issue for it, and fix it later. 
This PR fixes a real issue in code and it is causing usability issues at least with AVSystem.

We can convert the Comby script to Coccinelle later.

**Review by VeijoPesonen — DISMISSED** [review]

**Comment by galak** [comment]
I think if the files are committed into the tree they should conform to the coding style.

**Comment by broglep-work** [comment]
> I think if the files are committed into the tree they should conform to the coding style.

I generally agree, the underlying issue here however is we have conflicting style guidelines, I opened #52712 for this. So the only way to resolve this at the moment is break one of the guidelines.

**Comment by SeppoTakalo** [comment]
> I think if the files are committed into the tree they should conform to the coding style.

Maybe this rule should be understood that files, we expect humans to touch.

After all, we are now talking about generated parser which we should not allow anyone to directly edit.

**Review by SeppoTakalo — APPROVED** [review]

**Comment by carlescufi** [comment]
@broglep-work during last week's TSC there seemed to be clear agreement as to not allow yet another (i.e. comby) source code parser/reformatter. I suggest the following as alternatives:

- Commit the untouched files, then immediately commit the modified (by hand or coccinelle) files so we keep track of the diff with Git
- Commit the modified files and a coccinelle script
- Commit the modified files and a `.patch`file that diffs the generatd vs modified files

Any of those should work, but please remove the .comby stuff.

File: .checkpatch.conf [hunk 2]
```diff
@@ -29,3 +29,4 @@
 --ignore MULTISTATEMENT_MACRO_USE_DO_WHILE
 --ignore ENOSYS
 --ignore IS_ENABLED_CONFIG
+--exclude subsys/net/lib/lwm2m/generated
```
Comment (carlescufi) [inline .checkpatch.conf:32]: remove this

File: subsys/net/lib/lwm2m/codegen/lwm2m_senml_cbor_comby.conf [hunk 3]
```diff
@@ -0,0 +1,120 @@
+#
```
Comment (carlescufi) [inline subsys/net/lib/lwm2m/codegen/lwm2m_senml_cbor_comby.conf:1]: remove this file

**Review by carlescufi — CHANGES_REQUESTED** [review]
see my comment with proposed solutions

**Comment by broglep-work** [comment]
> @broglep-work during last week's TSC there seemed to be clear agreement as to not allow yet another (i.e. comby) source code parser/reformatter. I suggest the following as alternatives:
> 
> * Commit the untouched files, then immediately commit the modified (by hand or coccinelle) files so we keep track of the diff with Git
> * Commit the modified files and a coccinelle script
> * Commit the modified files and a `.patch`file that diffs the generatd vs modified files
> 
> Any of those should work, but please remove the .comby stuff.

I will have to revisit this at a later point in time as I had to move on to other topics in the meantime. 

Based on what I've learned, probably only option 1) will be possible, after maybe performing some tool assisted modification and after that, tweak the code manually to pass code styles (as there is no guarantee that clang-format produces correct style, and the expectation is that the developer after that manually adjustes the code, see #52712) 
While this works, it is not really nice developer experience, and the next developer that needs touch that code needs to bite the bullet.

Option 2) might have the issue that coccinelle might not be able to perform formatting changes ( e.g break lines to adhere to coding style). So a single coccinelle transformation might not be sufficient and some additional tool is required (or manually changes). But as I mentioned, I'm not familiar with that tool, maybe it is also capable of transforming the code to adhere to style guides with some human help where clang-format itself fails.
Option 3) Using patch files has the issue that it can trigger checkpatch to think it is c code and complain about it (which happened in this PR). Same risk probably also exists with Option 2). It also is not really good in encoding the intended modification and not robust enough that the patch can be applied for modified / regenerated code. So also not of much use. 

The only option that seems valuable to me and worth investing time is removing the generated code from tree and generate it on-the-fly at compile time. This will resolve probably all problems with the current solution.

**Comment by carlescufi** [comment]
> The only option that seems valuable to me and worth investing time is removing the generated code from tree and generate it on-the-fly at compile time. This will resolve probably all problems with the current solution.

I understand but at the moment there is no consensus regarding code generation at build/generation time in Zephyr as a project, especially in cases like this, so this will require some discussion and may take a while.

**Comment by rlubos** [comment]
@broglep-work Would you mind if we took your commits and open a separate PR, w/o introducing new tools (comby) to get the bugfix merged? We'd like to move things forward.

**Comment by SeppoTakalo** [comment]
I have assigned this PR to our internal backlog, so we will refactor this PR into a separate PR without introducing dependency to a new tool.
After all, the bugfix is valid and the CDDL code is already provided so the hard work is already done.

**Comment by broglep-work** [comment]
I am back to work after the holiday break, and found this a nice christmas / new year gift, thanks for moving this forward in my absence

**Comment by jheiskan81** [comment]
@broglep-work Thanks for good contribute which was easy to push forward.

**Pull request closed without merging**